File: quicksort.py
# Inplace Quicksort for Pandas DataFrames
#   Based on code by Mohit Kumra and Anush Krishna V


# Swap two rows in a DataFrame

def swap_rows(df, i, j):
    df.iloc[i], df.iloc[j] = df.iloc[j], df.iloc[i].copy()


# This function takes last element as pivot, places
# the pivot element at its correct position in sorted
# array, and places all smaller (smaller than pivot)
# to left of pivot and all greater elements to right
# of pivot

def partition(df, sort_col, low, high):
    i = (low-1)             # index of smaller element
    arr = df[sort_col]      # col to be sorted

    pivot = arr.iloc[high]  # pivot

    for j in range(low, high):

        # If current element is smaller than or
        # equal to pivot
        if arr.iloc[j] <= pivot:

            # increment index of smaller element
            i = i+1
            swap_rows(df, i, j)

    swap_rows(df, i+1, high)
    return (i+1)


# The main function that implements QuickSort
# arr[] --> Array to be sorted,
# low  --> Starting index,
# high  --> Ending index

def quickSort(df, sort_col, low, high):
    if len(df) == 1:
        return df

    if low < high:
        # pi is partitioning index, arr[p] is now
        # at right place
        pi = partition(df, sort_col, low, high)

        # Separately sort elements before
        # partition and after partition
        quickSort(df, sort_col, low, pi-1)
        quickSort(df, sort_col, pi+1, high)

    return df

File: quicksort.jl
using Arrow, DataFrames
using PyCall

function swap_rows!(df::DataFrame, i::Int, j::Int)::Nothing
    df[i,:], df[j,:] = df[j,:], copy(df[i,:])
    return nothing
end

function partition(df::DataFrame,
                   sort_col::String,
                   low::Int,
                   high::Int
                   )::Int

    i = low
    arr = df[!, sort_col]
    pivot = arr[high]

    for j in low:high-1
        if arr[j] <= pivot
            swap_rows!(df, i, j)
            i += 1
        end
    end

    swap_rows!(df, i, high)
    return i
end

function quickSort!(df::DataFrame,
                    sort_col::String,
                    low::Int,
                    high::Int
                    )::DataFrame
    if size(df, 1) == 1
        return df
    end

    if low < high
        pi = partition(df, sort_col, low, high)

        quickSort!(df, sort_col, low, pi-1)
        quickSort!(df, sort_col, pi+1, high)
    end
    return df
end

function quickSort(bytes::Vector{UInt8},
                   sort_col::String,
                   low::Int,
                   high::Int
                   )::PyObject

    df = DataFrame(Arrow.Table(bytes))

    result = quickSort!(df, sort_col, low, high)

    io = IOBuffer()
    Arrow.write(io, result)
    seekstart(io)

    result_bytes = take!(io)
    return PyCall.pybytes(result_bytes)
end
